File: LeetCodeProblemsCMake/Problems/LeetCode1390FourDivisors.cpp
//
// Created by saeed on 1/4/26.
//

#include "LeetCode1390FourDivisors.h"

#include <complex>

int LeetCode1390FourDivisors::sumFourDivisors(std::vector<int> &nums) {
    int divisorCount{},divisorSum{},totalSum{},divisorCap{},otherDivisor{};
    for (auto num:nums) {
        divisorCount=1;
        divisorSum=1;
        divisorCap= static_cast<int>( std::sqrt(num));

        for ( int i=2;i<=divisorCap;++i) {
            if (num%i!=0) {
                continue;
            }
            divisorSum+=i;
            ++divisorCount;
            otherDivisor=num/i;
            if (otherDivisor!=i) {
                divisorSum+=otherDivisor;
                ++divisorCount;
            }
            if (divisorCount>3)
                break;
        }
        if (divisorCount==3) {
            totalSum+=divisorSum+num;
        }
    }
    return totalSum;
}

File: LeetCodeProblemsGo/LeetCode1390FourDivisors/LeetCode1390FourDivisors.go
package LeetCode1390FourDivisors

import "math"

//	func sumFourDivisors(nums []int) int {
//		totalSum := 0
//		divisorSum := 0
//		divisorCount := 0
//		for _, num := range nums {
//			divisorSum = 1
//			divisorCount = 1
//			for i := 2; i <= num/2; i++ {
//				if num%i == 0 {
//					divisorSum += i
//					divisorCount++
//					if divisorCount > 3 {
//						break
//					}
//				}
//			}
//			if divisorCount == 3 {
//				totalSum += divisorSum + num
//			}
//		}
//		return totalSum
//	}
func sumFourDivisors(nums []int) int {
	totalSum := 0
	divisorSum := 0
	divisorCount := 0
	divisorCap := 0
	otherDivisor := 0
	for _, num := range nums {
		divisorSum = 1
		divisorCount = 1
		divisorCap = int(math.Sqrt(float64(num)))
		for i := 2; i <= divisorCap; i++ {
			if num%i == 0 {
				divisorSum += i
				divisorCount++
				otherDivisor = num / i
				if otherDivisor != i {
					divisorSum += otherDivisor
					divisorCount++
				}
				if divisorCount > 3 {
					break
				}
			}
		}
		if divisorCount == 3 {
			totalSum += divisorSum + num
		}
	}
	return totalSum
}

func Calculate() {
	nums := []int{21, 4, 7}
	sum := sumFourDivisors(nums)
	println(sum)
}
